### ISSUE TITLE
Compressed Word-list Causes Windows Agent Unhandled Exception

### ISSUE THREAD

**Issue by rturowski** [s3inlc/hashtopussy#222]
#### Before you submit an issue please include the following information if you do not your issue will be closed.

Your current Server version located at the bottom of any hashtopussy webpage.
Hashtopussy: 0.3.2

Current Client version
Client Version 0.43.13, running on Windows 10 update 1703 x64

Your current Hashcat version
3.6.0

The exact task command you are trying to run.
\#HL# rockyou.txt -r _NSAKEY.v2.dive.rule

Debug output from the client by running "hashtopussy.exe debug"
D:\hashtopussy>hashtopussy.exe debug
Client Version 0.43.13
Checking for client updates
{"action":"update","type":"csharp","version":"0.43.13"}
{"action":"update","response":"SUCCESS","version":"OK"}
You are using the latest client version
System is Windows
Existing token found
{"action":"login","token":"GPOmMogONV"}
{"action":"login","response":"SUCCESS","timeout":"30"}
Logged in to server
{"action":"download","type":"hashcat","token":"GPOmMogONV","force":0}
Hashcat version v3.6.0 found
Getting task
{"action":"task","token":"GPOmMogONV"}
{"action":"task","response":"SUCCESS","task":7,"attackcmd":"#HL# rockyou.??? -r _NSAKEY.v2.dive.rule","cmdpars":"--opencl-devices 1,2 --hash-type=5500","hashlist":2,"bench":30,"statustimer":5,"files":["_NSAKEY.v2.dive.rule","rockyou.7z"],"benchType":"run","hashlistAlias":"#HL#"}
Downloading hashlist for this task, please wait...
{"action":"hashes","token":"GPOmMogONV","hashlist":2}
Getting chunk...
{"action":"chunk","token":"GPOmMogONV","taskId":7}
{"action":"chunk","response":"SUCCESS","status":"keyspace_required"}
Server has requested the client to measure the keyspace for this task
Something went wrong with keyspace measuring

Unhandled Exception: System.FormatException: Input string was not in a correct format.
   at System.Number.StringToNumber(String str, NumberStyles options, NumberBuffer& number, NumberFormatInfo info, Boolean parseDecimal)
   at System.Number.ParseInt64(String value, NumberStyles options, NumberFormatInfo numfmt)
   at System.Convert.ToInt64(String value)
   at hashtopussy.hashcatClass.parseKeyspace(String line, Int64& keySpace)
   at hashtopussy.hashcatClass.runKeyspace(Int64& keySpace)
   at hashtopussy.taskClass.getChunk(Int32 inTask)
   at hashtopussy.taskClass.getTask()
   at hashtopussy.Program.Main(String[] args)

Describe your problem in as much detail as possible " It's broke " is not a description.
Running the agent results in a crash when calculating the key space. In my case the problem only occurs when using a compressed word list (rockyou.7z). When the file is uncompressed, it leaves a stub file rockyou.7z which Windows incorrectly returns to the agent-issued command that looks for the wildcard "rockyou.???" . Here is my diagnostic command output with comments added:

Extracting the command line using SysInternals Process Monitor and executing it with the relative path added, shows the output of the issued hashcat command is help text.
```
D:\hashtopussy>"D:\hashtopussy\hashcat\hashcat64.exe"  --opencl-devices 1,2 --hash-type=5500  files\rockyou.??? -r files\_NSAKEY.v2.dive.rule --session=hashtopussy --keyspace --quiet
Usage: D:\hashtopussy\hashcat\hashcat64.exe [options]... hash|hashfile|hccapxfile [dictionary|mask|directory]...

Try --help for more help.
```
Rename the "rockyou.7z" file so that "rockyou.???" does not return it to hashcat.
```
D:\hashtopussy>ren files\rockyou.7z xrockyou.7z
```
Then the same hashcat command calculates and returns the keyspace as expected.
```
D:\hashtopussy>"D:\hashtopussy\hashcat\hashcat64.exe"  --opencl-devices 1,2 --hash-type=5500  files\rockyou.??? -r files\_NSAKEY.v2.dive.rule --session=hashtopussy --keyspace --quiet
14343296
```
Rename the "rockyou.7z" file back to the original name.
```
D:\hashtopussy>ren files\xrockyou.7z rockyou.7z
```
Run the same command and once again get the help message returned from hashcat.
```
D:\hashtopussy>"D:\hashtopussy\hashcat\hashcat64.exe"  --opencl-devices 1,2 --hash-type=5500  files\rockyou.??? -r files\_NSAKEY.v2.dive.rule --session=hashtopussy --keyspace --quiet
Usage: D:\hashtopussy\hashcat\hashcat64.exe [options]... hash|hashfile|hccapxfile [dictionary|mask|directory]...

Try --help for more help.
```
I believe this is root cause of why the agent crashes, at least in my case:
The Command Prompt DIR command (incorrectly, in my view) returns the 7z file and the txt file when using the wildcard extension ???.
```
D:\hashtopussy>dir files\rockyou.???
 Volume in drive D is Data
 Volume Serial Number is 5C81-B657

 Directory of D:\hashtopussy\files

07/02/2017  07:07 PM                 8 rockyou.7z
07/01/2017  09:17 PM       139,921,497 rockyou.txt
               2 File(s)    139,921,505 bytes
               0 Dir(s)  1,036,803,452,928 bytes free
```
PowerShell does it right, like Linux, and does not return the 7z file.
```
PS D:\hashtopussy> ls .\files\rockyou.???


    Directory: D:\hashtopussy\files


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----         7/1/2017   9:17 PM      139921497 rockyou.txt
```
However, hashtopussy agent did not seem to run correctly under PowerShell. Once I uploaded an uncompressed rockyou.txt file to use in my task, everything worked as expected on two different machines. I need to run Windows on most machines because of other requirements, so It would be nice to have a fix for this in code.

Thanks!

**Comment by winxp5421** [comment]
First off thanks for the detailed report, not many people do this. I thank you for that!

As far as your issue goes based on the debug output "attackcmd":"#HL# rockyou.??? -r _NSAKEY.v2.dive.rule" I believe you have some confusion about that ".???" means. You state the ??? as a wildcard it is not a wildcard it is simply a placeholder extension that you should change to the correct extension of the file located inside the 7z archive when you create your task within the HTP webUI. 

Proper procedure is as follows: (i will use rock you as an example for a compressed wordlist)
rockyou.txt >compress without any subfolders > rockyou.7z
upload rockyou.7z to HTP as a wordlist
create task> select rockyou.7z (rockyou.??? should appear) > manually change .??? to the correct .txt extension.
Profit.

The client handles this situation as follows:
Get task
Task has files to download
Check if files are already downloaded in ./files/ dir (they are not)
Download files: rockyou.7z ...
check if the file is compressed (yes it is)
Extract archive (creates ./files/rockyou.txt)
(Now there are 2 files one compressed and another uncompressed wasting space)
Client deletes compressed archive (rockyou.7z)
"type "UNCOMPRESSED" >rockyou.7z" or "cat "UNCOMPRESSED" > rockyou.7z"
Because the client ( and Hashtopussy) does not know what file is inside rockyou.7z there needs to be a placeholder or dummy file to take the archives place so the client knows this file was already downloaded from the server as the server and client are only aware of the file that was downloaded (rockyou.7z) not the file inside the archive.

TLDR;
.??? extension is not a wild card this needs to be changed manually when you create a task with a compressed wordlist if the file rockyou.txt is the file within the archive then the extension needs to be changed to .txt to match otherwise the client will not know what file you are talking about.

**Comment by rturowski** [comment]
Thank you for the explanation. It makes perfect sense now! :)  I guess I was always looking into it after the task failed, and I could not change the command at that time. Thanks for the great tool, and the installation videos too!
